runners/core-java/src/main/java/org/apache/beam/runners/core/UnboundedReadFromBoundedSource.java

```diff
@@ -108,11 +116,19 @@ public void validate() {
     public List<BoundedToUnboundedSourceAdapter<T>> generateInitialSplits(
         int desiredNumSplits, PipelineOptions options) throws Exception {
       long desiredBundleSize = boundedSource.getEstimatedSizeBytes(options) / desiredNumSplits;
-      List<BoundedToUnboundedSourceAdapter<T>> result = Lists.newArrayList();
-      for (BoundedSource<T> split : boundedSource.splitIntoBundles(desiredBundleSize, options)) {
-        result.add(new BoundedToUnboundedSourceAdapter<>(split));
+      List<? extends BoundedSource<T>> splits;
+      if (desiredBundleSize > 0
+          && (splits = boundedSource.splitIntoBundles(desiredBundleSize, options)) != null) {
+        return Lists.transform(
+            splits,
+            new Function<BoundedSource<T>, BoundedToUnboundedSourceAdapter<T>>() {
+              @Override
+              public BoundedToUnboundedSourceAdapter<T> apply(BoundedSource<T> input) {
+                return new BoundedToUnboundedSourceAdapter<>(input);
+              }});
+      } else {
```

Comment:
Handle if the split throws as well (and presumably fall back the same way).

When you fall back to `ImmutableList.of(this)`, log a warning that will explain to the user why this happened, able to distinguish the three causes.

* estimated size <= 0
* splitIntoBundles returned `null` (not legal return value).
* exception thrown by splitIntoBundles.